File: src/main/java/net/pterodactylus/sone/web/ajax/DeleteReplyAjaxPage.java
/*
 * Sone - DeleteReplyAjaxPage.java - Copyright © 2010–2016 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.web.ajax;

import com.google.common.base.Optional;

import net.pterodactylus.sone.data.PostReply;
import net.pterodactylus.sone.web.WebInterface;
import net.pterodactylus.sone.web.page.FreenetRequest;

/**
 * This AJAX page deletes a reply.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class DeleteReplyAjaxPage extends JsonPage {

	/**
	 * Creates a new AJAX page that deletes a reply.
	 *
	 * @param webInterface
	 *            The Sone web interface
	 */
	public DeleteReplyAjaxPage(WebInterface webInterface) {
		super("deleteReply.ajax", webInterface);
	}

	//
	// JSONPAGE METHODS
	//

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected JsonReturnObject createJsonObject(FreenetRequest request) {
		String replyId = request.getHttpRequest().getParam("reply");
		Optional<PostReply> reply = webInterface.getCore().getPostReply(replyId);
		if (!reply.isPresent()) {
			return createErrorJsonObject("invalid-reply-id");
		}
		if (!reply.get().getSone().isLocal()) {
			return createErrorJsonObject("not-authorized");
		}
		webInterface.getCore().deleteReply(reply.get());
		return createSuccessJsonObject();
	}

}

File: src/main/kotlin/net/pterodactylus/sone/web/ajax/DeleteReplyAjaxPage.kt
package net.pterodactylus.sone.web.ajax

import net.pterodactylus.sone.data.*
import net.pterodactylus.sone.utils.*
import net.pterodactylus.sone.web.*
import net.pterodactylus.sone.web.page.*
import javax.inject.*

/**
 * This AJAX page deletes a reply.
 */
class DeleteReplyAjaxPage @Inject constructor(webInterface: WebInterface) :
		LoggedInJsonPage("deleteReply.ajax", webInterface) {

	override fun createJsonObject(currentSone: Sone, request: FreenetRequest) =
			request.parameters["reply"]
					?.let(core::getPostReply)
					?.let { reply ->
						reply.sone.isLocal.ifTrue {
							createSuccessJsonObject().also {
								core.deleteReply(reply)
							}
						} ?: createErrorJsonObject("not-authorized")
					} ?: createErrorJsonObject("invalid-reply-id")

}
